File: 14-state-management/02-react-mobx/src/types/todo.ts
/**
 * Todo 項目的類型定義
 */
export interface Todo {
  id: string;
  text: string;
  completed: boolean;
  createdAt: number;
}

/**
 * 過濾器類型
 */
export type FilterType = 'all' | 'active' | 'completed';

/**
 * Todo 輸入數據
 */
export interface TodoInput {
  text: string;
}

File: 10-backend-apis/08-ruby-rails/app/models/todo.rb
class Todo < ApplicationRecord
  # Validations
  validates :text, presence: true, length: { minimum: 1, maximum: 500 }
  validates :completed, inclusion: { in: [true, false] }

  # Sanitize text before saving to prevent XSS
  before_validation :sanitize_text

  # Scopes
  scope :active, -> { where(completed: false) }
  scope :completed, -> { where(completed: true) }
  scope :recent, -> { order(created_at: :desc) }

  # Default values
  after_initialize :set_defaults, if: :new_record?

  private

  def set_defaults
    self.completed ||= false
  end

  def sanitize_text
    require 'cgi'
    self.text = CGI.escapeHTML(text) if text.present?
  end
end
